eCommerce platforms/Adobe Magento/src/code/local/Microsoft/Dfp/Model/Observer.php

```diff
@@ -0,0 +1,251 @@
+						case 'refunded':
+							$this->dfp->log("  -----------   Invoking Refund API - START --------------  ");
+							$this->invokeRefundDFP($order);
+							$this->dfp->log("  -----------   Invoking Refund API - END --------------  ");
+							break;
+						case 'chargeback':
+							$this->dfp->log("  -----------   Invoking Chargeback API - START --------------  ");
+							$this->invokeChargebackDFP($order);
+							$this->dfp->log("  -----------   Invoking Chargeback API - END --------------  ");
+							break;
+					}
+				}
+			}
+		}
+
+		$this->dfp->log("  -----------  UpdateStatusToDFP Observer Event END --------------  ");
+	}
+
+	private function invokeRefundDFP($refund)
+	{
+		$payload = $this->generateRefundPayload($refund);
+		return $this->dfp->sendRefund($payload);
+	}
+
+	private function invokeChargebackDFP($order)
+	{
+		$payload = $this->generateChargebackPayload($order);
+		return $this->dfp->sendChargeback($payload);
+	}
+
+	private function invokePurchaseStatusDFP($order)
+	{
+		$payload = $this->generatePurchaseStatusPayload($order);
+		return $this->dfp->sendPurchaseStatus($payload);
+	}
+
+	private function invokeUpdateAccountDFP($customer)
+	{
+		$payload = $this->generateUpdateAccountPayload($customer);
+		return $this->dfp->sendUpdateAccount($payload);
+	}
+
+	private function generatePurchaseStatusPayload($order)
+	{
+		$_metadata = array(
+			"trackingId"		=> $this->dfp->GUID(),
+			"merchantTimeStamp"	=> date('c')
+		);
+
+		$status = $order->getData()['status'];
+		if ($status == 'order_approved') {
+			$status = 'Approved';
+		} else {
+			$status = 'Canceled';
+		}
+
+		$purchaseStatusPayload = array(
+			'purchaseId'	=> $order->getData()['increment_id'],
+			'statusType'	=> $status,
+			'statusDate'	=> date('c'),
+			'reason'	    => '',
+			'_metadata'		=> $_metadata
+
+		);
+		return $purchaseStatusPayload;
+	}
+
+	private function generateRefundPayload($refund)
+	{
+		$_metadata = array(
+			"trackingId"		=> $this->dfp->GUID(),
+			"merchantTimeStamp"	=> date('c')
+		);
+
+		$refundPayload = array(
+			'refundId'       	    => $this->dfp->GUID(),
+			'status'        		=> "Approved",
+			'bankEventTimestamp'	=> date('c'),
+			'amount'     			=> $refund->getGrandTotal(),
+			'currency'     	        => $refund->getOrderCurrencyCode(),
+			'userId'				=> $refund->getCustomerEmail(),
+			'purchaseId'			=> $refund->getIncrementId(),
+			'_metadata'     		=> $_metadata
+		);
+		return $refundPayload;
+	}
+
+	private function generateUpdateAccountPayload($customer)
+	{
+		$_metadata = array(
+			"trackingId"		=> $this->dfp->GUID(),
+			"merchantTimeStamp"	=> date('c')
+		);
+
+		// Device FingerPrinting Details
+		$deviceContext = array(
+			"deviceContextId"	=> Mage::getSingleton('core/session')->getFptDfpSessionId(),
+			"ipAddress"			=> getenv('HTTP_CLIENT_IP') ?: getenv('HTTP_X_FORWARDED_FOR') ?: getenv('HTTP_X_FORWARDED') ?: getenv('HTTP_FORWARDED_FOR') ?: getenv('HTTP_FORWARDED') ?: getenv('REMOTE_ADDR'),
```

Comment:
It looks like ?: is actually a ternary with the true condition implied instead of a true null coalescing operator. Any reason these shouldn't be the actual ?? null coalescing operators?

Reply:
Good call out!  That's what I get for assuming.  I've changed all to ??